File: sample/javaSample/src/main/java/jp/co/geniee/samples/gnad/banner/MultipleBannersListViewAdapter.java
package jp.co.geniee.samples.gnad.banner;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import jp.co.geniee.samples.R;
import jp.co.geniee.samples.gnad.common.CellData;
import jp.co.geniee.samples.gnad.common.ImageGetTask;
import jp.co.geniee.gnadsdk.banner.GNAdView;

class MultipleBannersListViewAdapter extends ArrayAdapter<Object> {
    private static final int TYPE_ITEM = 0;
    private static final int TYPE_AD = 1;
    private static final int TYPE_MAX_COUNT = TYPE_AD + 1;

    private LayoutInflater inflater;

    private ArrayList<Object> dataList = new ArrayList<>();

    public MultipleBannersListViewAdapter(Context context, List<Object> list) {
        super(context, 0, list);
        this.inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        this.dataList = (ArrayList<Object>) list;
    }

    @Override
    public int getItemViewType(int position) {
        return (dataList.get(position) instanceof GNAdView) ? TYPE_AD : TYPE_ITEM;
    }

    @Override
    public int getViewTypeCount() {
        return TYPE_MAX_COUNT;
    }

    @Override
    public int getCount() {
        return dataList.size();
    }

    @Override
    public Object getItem(int position) {
        return dataList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    class ViewHolder {
        ImageView imageView;
        TextView textView;
        TextView tvDescription;
        LinearLayout adLayout;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder holder;
        int type = getItemViewType(position);
        if (convertView == null) {
            holder = new ViewHolder();
            switch (type) {
                case TYPE_AD:
                    convertView = inflater.inflate(R.layout.ad_item, null);
                    holder.adLayout = convertView.findViewById(R.id.ad_layout);
                    break;
                case TYPE_ITEM:
                    convertView = inflater.inflate(R.layout.listview_row_item, null);
                    holder.textView = convertView.findViewById(R.id.textView);
                    holder.imageView = convertView.findViewById(R.id.imageView);
                    holder.tvDescription = convertView.findViewById(R.id.tvDescription);
                    break;
            }
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

        Object cell = getItem(position);
        // Rendering SDK GNAdView content
        if (cell instanceof GNAdView) {
            holder.adLayout.removeAllViews();
            GNAdView adView = (GNAdView)cell;
            if (adView.getParent() == null) {

                holder.adLayout.addView(adView, LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);
                adView.showBannerWithSize(300, 100);
                adView.startAdLoop();
            }
        } else if(cell instanceof CellData){

            holder.textView.setText(((CellData) cell).title);
            holder.tvDescription.setText(((CellData) cell).content);
            holder.imageView.setTag(((CellData) cell).imgURL);
            new ImageGetTask(holder.imageView, ((CellData) cell).imgURL).execute();
        }
        return convertView;
    }
}

File: sample/kotlinSample/src/main/java/jp/co/geniee/samples/gnad/banner/MultipleBannersListViewAdapter.kt
package jp.co.geniee.samples.gnad.banner

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import jp.co.geniee.samples.R
import jp.co.geniee.samples.gnad.common.CellData
import jp.co.geniee.samples.gnad.common.ImageGetTask
import jp.co.geniee.gnadsdk.banner.GNAdView
import java.util.*

internal class MultipleBannersListViewAdapter(context: Context, list: List<Any>) : ArrayAdapter<Any>(context, 0, list) {

    private val inflater: LayoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    private var dataList = ArrayList<Any>()

    init {

        this.dataList = list as ArrayList<Any>
    }

    override fun getItemViewType(position: Int): Int {
        return if (dataList[position] is GNAdView) TYPE_AD else TYPE_ITEM
    }

    override fun getViewTypeCount(): Int {
        return TYPE_MAX_COUNT
    }

    override fun getCount(): Int {
        return dataList.size
    }

    override fun getItem(position: Int): Any? {
        return dataList[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    internal inner class ViewHolder {
        var imageView: ImageView? = null
        var textView: TextView? = null
        var tvDescription: TextView? = null
        var adLayout: LinearLayout? = null
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        val holder: ViewHolder
        val type = getItemViewType(position)
        if (view == null) {
            holder = ViewHolder()
            when (type) {
                TYPE_AD -> {
                    view = inflater.inflate(R.layout.ad_item, null)
                    holder.adLayout = view!!.findViewById(R.id.ad_layout)
                }
                TYPE_ITEM -> {
                    view = inflater.inflate(R.layout.listview_row_item, null)
                    holder.textView = view!!.findViewById(R.id.textView)
                    holder.imageView = view.findViewById(R.id.imageView)
                    holder.tvDescription = view.findViewById(R.id.tvDescription)
                }
            }
            view!!.tag = holder
        } else {
            holder = view.tag as ViewHolder
        }

        val cell = getItem(position)
        // Rendering SDK GNAdView content
        if (cell is GNAdView) {
            holder.adLayout!!.removeAllViews()
            if (cell.parent == null) {
                holder.adLayout!!.addView(cell, LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT)
                cell.showBannerWithSize(300, 100)
                cell.startAdLoop()
            }

        } else if (cell is CellData) {

            holder.textView!!.text = cell.title
            holder.tvDescription!!.text = cell.content
            holder.imageView!!.tag = cell.imgURL
            ImageGetTask(holder.imageView!!, cell.imgURL).execute()
        }
        return view
    }

    companion object {
        private val TYPE_ITEM = 0
        private val TYPE_AD = 1
        private val TYPE_MAX_COUNT = TYPE_AD + 1
    }
}
